File: src/main/java/me/itay/bluej/resourcelocation/BlueJResourceManager.java
package me.itay.bluej.resourcelocation;

import java.util.HashMap;

import me.itay.bluej.resourcelocation.builtin.file.BlueJFileSystemResolver;

public class BlueJResourceManager {
	
	private static HashMap<String, BlueJResolver> resolvers = new HashMap<>();
	
	static {
		registerResolver("file", new BlueJFileSystemResolver());
	}
	
	public static void registerResolver(String domain, BlueJResolver resolver) {
		if(resolvers.containsKey(domain)) {
			throw new IllegalArgumentException("resolver for that domain already exists");
		}
		resolvers.put(domain, resolver);
	}
	
	public static BlueJResolvedResource resolve(String path) {
		return resolve(new BlueJResourceLocation(path));
	}
	
	public static BlueJResolvedResource resolve(BlueJResourceLocation location) {
		if(resolvers.containsKey(location.getDomain())) {
			if(location.isBase()) {
				return resolvers.get(location.getDomain()).resolve(location, null);
			}
			BlueJResolvedResource base = resolve(location.getParent());
			return resolvers.get(location.getDomain()).resolve(location, base);
		}else {
			return null;
		}
	}
	
}

File: src/main/java/me/itay/bluej/resourcelocation/BlueJResourceManager.kt
package me.itay.bluej.resourcelocation

import me.itay.bluej.resourcelocation.builtin.file.BlueJFileSystemResolver
import java.util.HashMap

private val resolvers = HashMap<String, BlueJResolver>()

fun init(){
    registerResolver("file", BlueJFileSystemResolver())
}

fun registerResolver(domain: String, resolver: BlueJResolver) {
    if (resolvers.containsKey(domain)) {
        throw IllegalArgumentException("resolver for that domain already exists")
    }
    resolvers[domain] = resolver
}

fun resolve(path: String): BlueJResolvedResource? {
    return resolve(BlueJResourceLocation(path))
}

fun resolve(location: BlueJResourceLocation): BlueJResolvedResource? {
    if (resolvers.containsKey(location.domain)) {
        if (location.isBase) {
            return resolvers[location.domain]!!.resolve(location, null)
        }
        val base = resolve(location.parent)
        return resolvers[location.domain]!!.resolve(location, base)
    }
    return null
}